thrifttest/streamtest/mock_stream.go

```diff
@@ -0,0 +1,960 @@
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+// Source: go.uber.org/thriftrw/protocol/stream (interfaces: Protocol,Writer,Reader,BodyReader,Enveloper,RequestReader,ResponseWriter)
+
+// Package streamtest is a generated GoMock package.
+package streamtest
+
+import (
+	context "context"
+	gomock "github.com/golang/mock/gomock"
+	stream "go.uber.org/thriftrw/protocol/stream"
+	wire "go.uber.org/thriftrw/wire"
+	io "io"
+	reflect "reflect"
+)
+
+// MockProtocol is a mock of Protocol interface
+type MockProtocol struct {
+	ctrl     *gomock.Controller
+	recorder *MockProtocolMockRecorder
+}
+
+// MockProtocolMockRecorder is the mock recorder for MockProtocol
+type MockProtocolMockRecorder struct {
+	mock *MockProtocol
+}
+
+// NewMockProtocol creates a new mock instance
+func NewMockProtocol(ctrl *gomock.Controller) *MockProtocol {
+	mock := &MockProtocol{ctrl: ctrl}
+	mock.recorder = &MockProtocolMockRecorder{mock}
+	return mock
+}
+
+// EXPECT returns an object that allows the caller to indicate expected use
+func (m *MockProtocol) EXPECT() *MockProtocolMockRecorder {
+	return m.recorder
+}
+
+// Reader mocks base method
+func (m *MockProtocol) Reader(arg0 io.Reader) stream.Reader {
+	m.ctrl.T.Helper()
+	ret := m.ctrl.Call(m, "Reader", arg0)
+	ret0, _ := ret[0].(stream.Reader)
+	return ret0
+}
+
+// Reader indicates an expected call of Reader
+func (mr *MockProtocolMockRecorder) Reader(arg0 interface{}) *gomock.Call {
+	mr.mock.ctrl.T.Helper()
+	return mr.mock.ctrl.RecordCallWithMethodType(mr.mock, "Reader", reflect.TypeOf((*MockProtocol)(nil).Reader), arg0)
+}
+
+// Writer mocks base method
+func (m *MockProtocol) Writer(arg0 io.Writer) stream.Writer {
+	m.ctrl.T.Helper()
+	ret := m.ctrl.Call(m, "Writer", arg0)
+	ret0, _ := ret[0].(stream.Writer)
+	return ret0
+}
+
+// Writer indicates an expected call of Writer
+func (mr *MockProtocolMockRecorder) Writer(arg0 interface{}) *gomock.Call {
+	mr.mock.ctrl.T.Helper()
+	return mr.mock.ctrl.RecordCallWithMethodType(mr.mock, "Writer", reflect.TypeOf((*MockProtocol)(nil).Writer), arg0)
+}
+
+// private mocks base method
+func (m *MockProtocol) private() {
```

Comment:
So turns out this won't work. This MockProtocol does not implement the stream.Protocol interface because of the `private()` method.

Reply:
You're correct and I'm ahead of you on this :).

There're currently two branches that I'm using to dev from:
https://github.com/thriftrw/thriftrw-go/tree/witriew/unprivate_ifaces
https://github.com/thriftrw/thriftrw-go/tree/witriew/embed_private_mock

The branch names are sort of self-explanatory.  If the interfaces that we decided here look good enough for releasing, I'll use the former branch.  Otherwise, I'll use the latter branch to make them implement `iface.Private`, but this means that we'll need to do this for the mock generation each time.

Reply:
I created #532 to resolve this.